pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -103,206 +113,260 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
 		&api.Secret{},
 		options.SecretResync,
 		framework.ResourceEventHandlerFuncs{
-			AddFunc:    e.secretAdded,
-			UpdateFunc: e.secretUpdated,
-			DeleteFunc: e.secretDeleted,
+			AddFunc:    e.queueSecretSync,
+			UpdateFunc: e.queueSecretUpdateSync,
+			DeleteFunc: e.queueSecretSync,
 		},
 		cache.Indexers{"namespace": cache.MetaNamespaceIndexFunc},
 	)
 
-	e.serviceAccountsSynced = e.serviceAccountController.HasSynced
-	e.secretsSynced = e.secretController.HasSynced
-
 	return e
 }
 
 // TokensController manages ServiceAccountToken secrets for ServiceAccount objects
 type TokensController struct {
-	stopChan chan struct{}
-
 	client clientset.Interface
 	token  serviceaccount.TokenGenerator
 
 	rootCA []byte
 
-	serviceAccounts cache.Indexer
+	serviceAccounts cache.Store
 	secrets         cache.Indexer
 
 	// Since we join two objects, we'll watch both of them with controllers.
 	serviceAccountController *framework.Controller
 	secretController         *framework.Controller
 
-	// These are here so tests can inject a 'return true'.
-	serviceAccountsSynced func() bool
-	secretsSynced         func() bool
+	// syncServiceAccountQueue handles service account events:
+	//   * ensures a referenced token exists for service accounts which still exist
+	//   * ensures tokens are removed for service accounts which no longer exist
+	// key is "<namespace>/<name>/<uid>"
+	syncServiceAccountQueue *workqueue.Type
+
+	// syncSecretQueue handles secret events:
+	//   * deletes tokens whose service account no longer exists
+	//   * updates tokens with missing token or namespace data, or mismatched ca data
+	//   * ensures service account secret references are removed for tokens which are deleted
+	// key is "<namespace>/<name>/<uid>/<serviceaccount-name>/<serviceaccount-uid>" (serviceaccount-uid may be blank)
```

Comment:
Is it OK if `ns/n/uid/account-name/uid` and `ns/n/uid/account-name/` are processed concurrently?